futures-util/src/io/into_sink.rs

```diff
@@ -0,0 +1,114 @@
+use futures_core::task::{Context, Poll};
+use futures_io::AsyncWrite;
+use futures_sink::Sink;
+use std::fmt;
+use std::io;
+use std::pin::Pin;
+use std::marker::Unpin;
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+
+struct Block {
+    offset: usize,
+    bytes: Box<dyn AsRef<[u8]>>,
```

Comment:
I'd make `Block` and `IntoSink` generic on the kind of `Bytes: AsRef<[u8]>` they accept. This would allow folks to send in e.g. references to buffers, as sending vectors without double-allocating them. If folks want to send multiple different kinds of `Bytes`, they can set `Box<dyn AsRef[u8]>` as the generic parameter.

Note that this would allow dropping the `'static` bounds below.

Reply:
Done, it doesn't actually allow using `Box<dyn AsRef<[u8]>>` as there is `impl<T: ?Sized> AsRef<T> for Box<T>` instead of `impl<U, T: AsRef<U> + ?Sized> AsRef<U> for Box<T>`, but it should be possible to make a custom newtyped-`Box` with that impl if someone does need that behaviour.